### PR TITLE
feat: config plumbing for vector benchmark framework

### PR DESCRIPTION
**Pull request by chebbyChefNEQ**
a lance oriented vector benchmark framework.

This framework is similar in idea to ann-benchmark, with a lot more focus poured into exploring how index and query hyperparameters affect recall instead of just plotting the recall/throughput frontier.

### CODE REVIEW

File: benchmarks/full_report/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import pytest
+import tempfile
+import lance
+
+from vbench.utils import make_random_vector_table
+
+
+@pytest.fixture
+def temdir():
```
Comment (wjones127) [inline benchmarks/full_report/tests/conftest.py:23]: small nit: I'd prefer `tmp` or `temp` over `tem` prefix.

File: benchmarks/full_report/vbench/experiment.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+def _load_npy(path: str) -> np.ndarray:
+    return np.load(path)
+
+
+@attrs.define
+class AnnTest:
+    name: str = attrs.field()
+    query: np.ndarray = attrs.field(converter=_load_npy)
+    ground_truth: np.ndarray = attrs.field(converter=_load_npy)
```
Comment (wjones127) [inline benchmarks/full_report/vbench/experiment.py:33]: there a reason not to just pass the functions as is?

```suggestion
    query: np.ndarray = attrs.field(converter=np.load)
    ground_truth: np.ndarray = attrs.field(converter=np.load)
```

File: benchmarks/full_report/vbench/experiment.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def _load_dataset(url: str) -> lance.LanceDataset:
+    return lance.dataset(url)
+
+
+@attrs.define
+class Benchmark:
+    name: str = attrs.field()
+    data: lance.LanceDataset = attrs.field(converter=_load_dataset)
```
Comment (wjones127) [inline benchmarks/full_report/vbench/experiment.py:43]:
```suggestion
    data: lance.LanceDataset = attrs.field(converter=lance.dataset)
```

File: benchmarks/full_report/vbench/runner.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            futs = []
+
+            # warmup
+            for _ in range(100):
+                futs.append(
+                    executor.submit(
+                        functools.partial(
+                            _test,
+                            np.random.standard_normal(test_case.query.shape[1]),
+                            list(range(k_gt)),
```
Comment (wjones127) [inline benchmarks/full_report/vbench/runner.py:96]: I'm not sure I get this part. Is the dataset sorted by the ground truth distance to a known query vector?

Comment (chebbyChefNEQ) [reply benchmarks/full_report/vbench/runner.py:96]: the gt and query here are just dummies meant for warming up the threads and dataset. the result is discarded later

Comment (wjones127) [reply benchmarks/full_report/vbench/runner.py:96]: Okay that makes more sense.

**Review by wjones127 — APPROVED** [review]

File: benchmarks/full_report/vbench/experiment.py (hunk 3 above)
Comment (chebbyChefNEQ) [reply benchmarks/full_report/vbench/experiment.py:44]: added better type resolution for this converter

**Pull request merged**